org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/LinkResourceUtil.java

```diff
@@ -0,0 +1,127 @@
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IFolder;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IProjectDescription;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.jdt.core.IJavaProject;
+
+class LinkResourceUtil {
+
+	private static final IPath METADATA_FOLDER_PATH = ResourcesPlugin.getWorkspace().getRoot().getLocation().append(".projects");
+
+	private static boolean isNewer(Path file, Instant instant) throws CoreException {
+		try {
+			BasicFileAttributes attributes = Files.getFileAttributeView(file, BasicFileAttributeView.class).readAttributes();
+			return attributes.creationTime().toInstant().isAfter(instant);
+		} catch (IOException ex) {
+			throw new CoreException(Status.error(ex.getMessage(), ex));
+		}
+	}
+
+    /**
+     * Get the redirected path of the input path. The path will be redirected to
+     * the workspace's metadata folder ({@link LinkResourceUtil#METADATA_FOLDER_PATH}).
+     * @param projectName name of the project.
+     * @param path path needs to be redirected.
+     * @return the redirected path.
+     */
+    private static IPath getMetaDataFilePath(String projectName, IPath path) {
+        if (path.segmentCount() == 1) {
+            return METADATA_FOLDER_PATH.append(projectName).append(path);
+        }
+
+        String lastSegment = path.lastSegment();
+        if (IProjectDescription.DESCRIPTION_FILE_NAME.equals(lastSegment)) {
+            return METADATA_FOLDER_PATH.append(projectName).append(lastSegment);
+        }
+
+        return null;
+    }
+
+	private static void linkFolderIfNewer(IFolder settingsFolder, Instant instant) throws CoreException {
+		if (settingsFolder.isLinked()) {
+			return;
+		}
+		if (settingsFolder.exists() && !isNewer(settingsFolder.getLocation().toPath(), instant)) {
+			return;
+		}
+		if (!settingsFolder.exists()) {
+			// not existing yet, create link
+			File diskFolder = getMetaDataFilePath(settingsFolder.getProject().getName(), settingsFolder.getProjectRelativePath()).toFile();
+			diskFolder.mkdirs();
+			settingsFolder.createLink(diskFolder.toURI(), IResource.NONE, new NullProgressMonitor());
+		} else if (isNewer(settingsFolder.getLocation().toPath(), instant)) {
+			// already existing but not existing before import: move then link
+			File sourceFolder = settingsFolder.getLocation().toFile();
+			File targetFolder = getMetaDataFilePath(settingsFolder.getProject().getName(), settingsFolder.getProjectRelativePath()).toFile();
+			File parentTargetFolder = targetFolder.getParentFile();
+			if (!parentTargetFolder.isDirectory()) {
+				parentTargetFolder.mkdirs();
+			}
+			sourceFolder.renameTo(targetFolder);
```

Comment:
`renameTo` can remove original folder but not file on Windows, interesting...

Reply:
I think it may be because the file was already opened by the process or something like that. IIRC Windows used to not delete files that still are open for reading; while other OSs allow that. Since folders cannot be open, this constraint doesn't apply. Maybe the behavior became an important side effect and while Windows may now allow it, the `renameTo` method keeps the older behavior; while the new `Files` API could deal without this constraint.

Reply:
Makes sense!